server/CGameHandler.cpp

```diff
@@ -3668,7 +3668,9 @@ bool CGameHandler::transformInUndead(const IMarket *market, const CGHeroInstance
 	//resulting creature - bone dragons or skeletons
 	CreatureID resCreature = CreatureID::SKELETON;
 
-	if (s.hasBonusOfType(Bonus::DRAGON_NATURE)
+	if ((s.hasBonusOfType(Bonus::DRAGON_NATURE)
+			&& (s.getCreatureID() != CreatureID::BONE_DRAGON)
+			&& (s.getCreatureID() != CreatureID::GHOST_DRAGON))
```

Comment:
This should be more generic solution f.e. `!hasBonusOfType(Bonus::UNDEAD)`

Reply:
We already have Dracolich for that matter

Reply:
I do not see how is it much better - 1 less line of code? Current conditions make it fairly clear why they are here in the first place.

Reply:
Condition should be: any dragon|hydra to bone dragon but undead dragon to skeleton.